### ISSUE TITLE
Design bigtable::Client and bigtable::AdminClient to be thread-safe

### ISSUE THREAD

**Issue by coryan** [GoogleCloudPlatform/google-cloud-cpp#101]
The classes are not thread safe today.  We may decide that they will never be, if so, we need to document this.  Most likely we will want them to be thread-safe, but that will change their interfaces.

Find out when the classes change state, if ever, if they do, document how we will make them safe, implement that design.

**Comment by coryan** [comment]
For the record, I think implementing credential refresh should not be solved as part of this bug, that is being tracked in #137.  But we should keep that in mind.

Also, this is related to #34.

I am going to outline the design here, because it is going to require a series of PRs and I want to be able to give folks a context when they are evaluating the PR.

We are aiming towards an client class that can be used safely from multiple threads.  That means any objects or references returned by the class should not be destroyed by other threads.  We also need to support different behaviors for #137.  The [`bigtable::AdminClient`](https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/701602715e51b2fd6fc3ed277308e29efb7c567a/bigtable/admin/admin_client.h#L27) class is sort of part of the way there.  it is an abstract class, so the behavior would be configurable. But it returns a `::google::bigtable::admin::v2::BigtableTableAdmin::StubInterface&`, and the underlying object changes when we need to reconnect to the server.  There is no way to make that thread safe.

I propose making the `bigtable::AdminClient` and `bigtable::DataClient` (yes, I am changing the name) interfaces look like this (using `DataClient` in the example, but applies to both):

```C++
class DataClient {
public:
  virtual ~DataClient() = default;
  virtual std::string const& ProjectId() const = 0;
  virtual std::string const& InstanceId() const = 0;

  // Return the current stub, creating one if necessary.  May round robin across several stubs.
  virtual std::shared_ptr<::google::bigtable::v2::Bigtable::StubInterface> TableStub() = 0;

  // Handle a RPC completion, reset the channels and stubs if necessary.
  virtual void OnCompletion(grpc::Status const& status) = 0;
};
```

The interface is safe when using multiple threads, a thread that gets an stub will not crash even if the underlying channel is closed.  All that would happen is that the next RPC would fail.  The classes that use this class should fetch a new stub before every RPC (no caching), which would support round robing an do updates after a call.

The implementation of the class would do whatever locking is needed to update the stub.


### Alternatives Considered

We could expose all the RPC operations as thin wrappers in the `DataClient` class, for example (types ommitted):

```C++
class DataClient {
public:
  virtual void ReadRows(...) = 0;
  virtual void ApplyBulk(...) = 0;
};
```

These functions would just call the right RPC, they would not retry, they would not expose or provide a higher-level API.  That way we do not have to return a stub of any kind, and any updates or resets can be done internally.  But this is a lot of extra typing of thin wrappers, rather not do it.

**Comment by coryan** [comment]
I am going to start on the implementation.  This will be done in multiple steps:

- [ ] Break the cycle of dependency between `bigtable::Client` and `bigtable::Table`, the client will not create a `Table`, a table will be created using a `std::shared_ptr<Client>` parameter.
- [] Rename the `bigtable::Client` class to `bigtable::DataClient`.
- [] Move the `bigtable::DataClient` class to its own file.  Also rename `data.{h,cc}` to `table.{h,cc}`
- [] Add the `ProjectId` and `InstanceId` member functions.
- [] Add some helper functions to create a full instance name (`projects/blah/instances/bleh`) from a `bigtable::DataClient` and a `bigtable::AdminClient`.
- [] Add some helper functions to create a full table name from the components while we are at it.
- [] Return a `std::shared_ptr<>` for the stub in `bigtable::DataClient` and `bigtable::AdminClient`.   I think that will fix #34.
- [] Add proper locking to the implementation of each.  That will close this bug.

**Comment by coryan** [comment]
Checklist is all closed.  Closing this bug.
